File: src/main/java/io/github/gravetii/game/GameResult.java
package io.github.gravetii.game;

import io.github.gravetii.model.GridPoint;
import io.github.gravetii.model.WordResult;

import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public class GameResult {
  private final Map<String, WordResult> wordToResultMap = new LinkedHashMap<>();
  private int totalScore = 0;
  private GameQuality quality;

  public void put(String word, int score, List<GridPoint> seq) {
    if (!this.wordToResultMap.containsKey(word)) {
      this.wordToResultMap.put(word, new WordResult(word, score, seq));
      this.totalScore += score;
    }
  }

  public boolean exists(String word) {
    return this.wordToResultMap.containsKey(word);
  }

  public int getTotalScore() {
    return this.totalScore;
  }

  public int getPoints(String word) {
    return this.wordToResultMap.get(word).getScore();
  }

  public Map<String, WordResult> all() {
    return this.wordToResultMap;
  }

  public WordResult get(String word) {
    return this.wordToResultMap.get(word);
  }

  public void defineQuality() {
    this.quality =
        Arrays.stream(GameQuality.values())
            .dropWhile(x -> this.wordToResultMap.size() <= x.getMinCount())
            .findFirst()
            .orElse(GameQuality.LOW);
  }

  public GameQuality getQuality() {
    return this.quality;
  }

  @Override
  public String toString() {
    return "GameResult{"
        + "wordToResultMap="
        + wordToResultMap
        + ", totalScore="
        + totalScore
        + ", quality="
        + quality
        + '}';
  }
}

File: src/main/kotlin/io/github/gravetii/game/GameResult.kt
package io.github.gravetii.game

import io.github.gravetii.model.GridPoint
import io.github.gravetii.model.WordResult

class GameResult {

    private var wordToResultMap: MutableMap<String, WordResult> = mutableMapOf()
    var totalScore: Int = 0
    var quality: GameQuality = GameQuality.LOW

    fun put(word: String, score: Int, seq: List<GridPoint>) {
        if (!wordToResultMap.containsKey(word)) {
            wordToResultMap[word] = WordResult(word, score, seq)
            totalScore += score
        }
    }

    fun exists(word: String): Boolean = wordToResultMap.containsKey(word)

    fun getPoints(word: String): Int = wordToResultMap[word]!!.score

    fun all(): Map<String, WordResult> = wordToResultMap

    fun get(word: String): WordResult = wordToResultMap[word]!!

    fun defineQuality(): GameResult {
        quality = GameQuality.values()
            .dropWhile { wordToResultMap.size < it.minCount }
            .firstOrNull() ?: GameQuality.LOW
        return this
    }

}